[ML] API integration tests for APM latency correlation.

walterra opened the pull request:
## Summary

Follow up to #99905.

Fixes #105049 (updated how we pass on APM's params to the async search service)

Adds API integration tests for APM Latency Correlations code.

Writing the tests surfaced some glitches fixed as part of this PR:

- If the applied filters don't return any docs, we won't throw an error anymore. Instead, the async search service finishes early and just returns no results.
- If for whatever reason the async search service throws an error, it will also set its state now to `isRunning = false`.
- If the client triggers a request with a service ID we now make sure that async search service still exists. We throw an error if that service no longer exists. This avoids re-instantiating async search services when they've already finished or failed and for whatever reason a client triggers another request with the same ID.
- Refactored requests to reuse APM's own `getCorrelationsFilters()`. We now require `start/end` to be set and it will be converted from ISO (client side) to epochmillis (server side) to be more in line with APM's existing code. 
- The async search service now creates a simple internal log. This gets exposed via the API and we assert it using the API tests. In the future, we might also expose it in the UI to allow for better problem investigation for users and support.

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

walterra committed:
API integration tests for latency correlation.

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/)
* Commit: 96a1a6029cb635bc43c42b6c44fde01ab637833f
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/96a1a6029cb635bc43c42b6c44fde01ab637833f)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/execution/node/190/log)
* [Execute xpack-kibana-ciGroup1](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/execution/node/541/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/","number":"136857","commit":"96a1a6029cb635bc43c42b6c44fde01ab637833f"}]}
PIPELINE-->

walterra committed:
fix unit tests.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/)
* Commit: 48074a40d509ab90057428a0ee64a7cfb7401cf8
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/48074a40d509ab90057428a0ee64a7cfb7401cf8)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/execution/node/190/log)

### History
* :broken_heart: [Build #136857](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/) failed 96a1a6029cb635bc43c42b6c44fde01ab637833f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/","number":"136857","commit":"96a1a6029cb635bc43c42b6c44fde01ab637833f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/","number":"136944","commit":"48074a40d509ab90057428a0ee64a7cfb7401cf8"}]}
PIPELINE-->

walterra committed:
fix handling of params.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/)
* Commit: 8250dd07bf81f3c36ba3a27fa24388d4ca7c778b
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/8250dd07bf81f3c36ba3a27fa24388d4ca7c778b)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/execution/node/190/log)
* [Execute xpack-kibana-ciGroup1](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/execution/node/541/log)

### History
* :broken_heart: [Build #136944](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/) failed 48074a40d509ab90057428a0ee64a7cfb7401cf8
* :broken_heart: [Build #136857](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/) failed 96a1a6029cb635bc43c42b6c44fde01ab637833f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/","number":"136857","commit":"96a1a6029cb635bc43c42b6c44fde01ab637833f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/","number":"136944","commit":"48074a40d509ab90057428a0ee64a7cfb7401cf8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/","number":"136969","commit":"8250dd07bf81f3c36ba3a27fa24388d4ca7c778b"}]}
PIPELINE-->

smith commented on x-pack/plugins/apm/server/lib/search_strategies/correlations/query_histogram_rangesteps.ts:
```diff
@@ -41,25 +49,26 @@ export const fetchTransactionDurationHistogramRangesteps = async (
   if (resp.body.aggregations === undefined) {
     throw new Error(
-      'fetchTransactionDurationHistogramInterval failed, did not return aggregations.'
+      'fetchTransactionDurationHistogramRangesteps failed, did not return aggregations.'
```
Is it `rangeSteps` or `rangesteps`? The file name and function name and error messages don't agree.

walterra replied:
Good catch! I renamed everything to be consistent `rangeSteps` and `range_steps` in 509a02d.

smith reviewed (commented):
Looks ok, but are the test failures related to the changes here? Is this ready for review or is there more work needed on making the tests pass?

walterra committed:
fix unit tests.

walterra committed:
Merge branch 'master' into ml-apm-correlations-api-integration-tests

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/)
* Commit: 509a02d37039f42c8e8919cf63ae795d4cfa6e97
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/509a02d37039f42c8e8919cf63ae795d4cfa6e97)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup1](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/execution/node/544/log)
* [Execute xpack-kibana-ciGroup1](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/execution/node/943/log)

### History
* :broken_heart: [Build #136969](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/) failed 8250dd07bf81f3c36ba3a27fa24388d4ca7c778b
* :broken_heart: [Build #136944](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/) failed 48074a40d509ab90057428a0ee64a7cfb7401cf8
* :broken_heart: [Build #136857](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/) failed 96a1a6029cb635bc43c42b6c44fde01ab637833f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/","number":"136857","commit":"96a1a6029cb635bc43c42b6c44fde01ab637833f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/","number":"136944","commit":"48074a40d509ab90057428a0ee64a7cfb7401cf8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/","number":"136969","commit":"8250dd07bf81f3c36ba3a27fa24388d4ca7c778b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/","number":"137133","commit":"509a02d37039f42c8e8919cf63ae795d4cfa6e97"}]}
PIPELINE-->

walterra committed:
adjust api/unit test

walterra commented:
@smith Thanks for the review, I thought it was ready but I overlooked that the bugs I fixed as part of this PR caused the existing unit tests to fail. Fixed those in 509a02d. Also identified an issue with how we applied filters which should now be fixed but needed changing the assertions for API tests. I'll flip this to draft again until all tests pass.

walterra committed:
assert internal log.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/)
* Commit: 5161477a1bd201dda403d1c22db84521a9cb95e4
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/5161477a1bd201dda403d1c22db84521a9cb95e4)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/646f1d4076fd1ab2589c429f09966d4659be7b71) | [after](https://github.com/elastic/kibana/commit/5161477a1bd201dda403d1c22db84521a9cb95e4) | diff |
| --- | --- | --- | --- |
| `apm` | 4.3MB | 4.3MB | -4.0B |

### History
* :broken_heart: [Build #137133](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/) failed 509a02d37039f42c8e8919cf63ae795d4cfa6e97
* :broken_heart: [Build #136969](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/) failed 8250dd07bf81f3c36ba3a27fa24388d4ca7c778b
* :broken_heart: [Build #136944](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/) failed 48074a40d509ab90057428a0ee64a7cfb7401cf8
* :broken_heart: [Build #136857](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/) failed 96a1a6029cb635bc43c42b6c44fde01ab637833f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/","number":"136857","commit":"96a1a6029cb635bc43c42b6c44fde01ab637833f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/","number":"136944","commit":"48074a40d509ab90057428a0ee64a7cfb7401cf8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/","number":"136969","commit":"8250dd07bf81f3c36ba3a27fa24388d4ca7c778b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/","number":"137133","commit":"509a02d37039f42c8e8919cf63ae795d4cfa6e97"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/","number":"137210","commit":"5161477a1bd201dda403d1c22db84521a9cb95e4"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137246/)
* Commit: 987f44bf1cd095b4a2167c0411aaee26e867f207
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/987f44bf1cd095b4a2167c0411aaee26e867f207)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137246/execution/node/190/log)

### History
* :green_heart: [Build #137210](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/) succeeded 5161477a1bd201dda403d1c22db84521a9cb95e4
* :broken_heart: [Build #137133](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/) failed 509a02d37039f42c8e8919cf63ae795d4cfa6e97
* :broken_heart: [Build #136969](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/) failed 8250dd07bf81f3c36ba3a27fa24388d4ca7c778b
* :broken_heart: [Build #136944](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/) failed 48074a40d509ab90057428a0ee64a7cfb7401cf8
* :broken_heart: [Build #136857](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/) failed 96a1a6029cb635bc43c42b6c44fde01ab637833f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136857/","number":"136857","commit":"96a1a6029cb635bc43c42b6c44fde01ab637833f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/","number":"136944","commit":"48074a40d509ab90057428a0ee64a7cfb7401cf8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/","number":"136969","commit":"8250dd07bf81f3c36ba3a27fa24388d4ca7c778b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/","number":"137133","commit":"509a02d37039f42c8e8919cf63ae795d4cfa6e97"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/","number":"137210","commit":"5161477a1bd201dda403d1c22db84521a9cb95e4"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137246/","number":"137246","commit":"987f44bf1cd095b4a2167c0411aaee26e867f207"}]}
PIPELINE-->

walterra committed:
fix unit tests.

walterra committed:
Merge branch 'master' into ml-apm-correlations-api-integration-tests

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137451/)
* Commit: 2fc22730066e83fd6b98f842d7e71d2f39d48315
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/2fc22730066e83fd6b98f842d7e71d2f39d48315)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/646f1d4076fd1ab2589c429f09966d4659be7b71) | [after](https://github.com/elastic/kibana/commit/2fc22730066e83fd6b98f842d7e71d2f39d48315) | diff |
| --- | --- | --- | --- |
| `apm` | 4.3MB | 4.3MB | -4.0B |

### History
* :broken_heart: [Build #137246](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137246/) failed 987f44bf1cd095b4a2167c0411aaee26e867f207
* :green_heart: [Build #137210](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/) succeeded 5161477a1bd201dda403d1c22db84521a9cb95e4
* :broken_heart: [Build #137133](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/) failed 509a02d37039f42c8e8919cf63ae795d4cfa6e97
* :broken_heart: [Build #136969](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/) failed 8250dd07bf81f3c36ba3a27fa24388d4ca7c778b
* :broken_heart: [Build #136944](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/) failed 48074a40d509ab90057428a0ee64a7cfb7401cf8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136944/","number":"136944","commit":"48074a40d509ab90057428a0ee64a7cfb7401cf8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/","number":"136969","commit":"8250dd07bf81f3c36ba3a27fa24388d4ca7c778b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/","number":"137133","commit":"509a02d37039f42c8e8919cf63ae795d4cfa6e97"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/","number":"137210","commit":"5161477a1bd201dda403d1c22db84521a9cb95e4"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137246/","number":"137246","commit":"987f44bf1cd095b4a2167c0411aaee26e867f207"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137451/","number":"137451","commit":"2fc22730066e83fd6b98f842d7e71d2f39d48315"}]}
PIPELINE-->

walterra commented:
Tests passed, this is ready for another look! @smith @qn895

walterra committed:
Merge branch 'master' into ml-apm-correlations-api-integration-tests

qn895 commented:
Tested and LGTM 🎉

walterra committed:
dataset with correlations. fix how to return errors for correlation agg.

walterra committed:
Merge branch 'master' into ml-apm-correlations-api-integration-tests

walterra commented:
I switched the dataset used for the tests from `apm_8.0.0` to `ml_8.0.0`, this one is able to return an actual correlation based on the limited data.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137739/)
* Commit: 8b008692e911fe377aa5b7124053268985530ee8
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104644/8b008692e911fe377aa5b7124053268985530ee8)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b2b57a240423e9503d1307075d34c73eb8b2a3c9) | [after](https://github.com/elastic/kibana/commit/8b008692e911fe377aa5b7124053268985530ee8) | diff |
| --- | --- | --- | --- |
| `apm` | 4.3MB | 4.3MB | -4.0B |

### History
* :green_heart: [Build #137451](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137451/) succeeded 2fc22730066e83fd6b98f842d7e71d2f39d48315
* :broken_heart: [Build #137246](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137246/) failed 987f44bf1cd095b4a2167c0411aaee26e867f207
* :green_heart: [Build #137210](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/) succeeded 5161477a1bd201dda403d1c22db84521a9cb95e4
* :broken_heart: [Build #137133](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/) failed 509a02d37039f42c8e8919cf63ae795d4cfa6e97
* :broken_heart: [Build #136969](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/) failed 8250dd07bf81f3c36ba3a27fa24388d4ca7c778b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @walterra

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136969/","number":"136969","commit":"8250dd07bf81f3c36ba3a27fa24388d4ca7c778b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137133/","number":"137133","commit":"509a02d37039f42c8e8919cf63ae795d4cfa6e97"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137210/","number":"137210","commit":"5161477a1bd201dda403d1c22db84521a9cb95e4"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137246/","number":"137246","commit":"987f44bf1cd095b4a2167c0411aaee26e867f207"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137451/","number":"137451","commit":"2fc22730066e83fd6b98f842d7e71d2f39d48315"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137739/","number":"137739","commit":"8b008692e911fe377aa5b7124053268985530ee8"}]}
PIPELINE-->

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.14](https://github.com/elastic/kibana/pull/105506)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/105506">](https://github.com/elastic/kibana/pull/105506) |
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/105507)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/105507">](https://github.com/elastic/kibana/pull/105507) |

The backport PRs will be merged automatically after passing CI.
